### DISCUSSION TITLE
Image Background Removal -- Assistance tweaking Magick command

### DISCUSSION THREAD

**Discussion by gaiking-uk** [ImageMagick/ImageMagick#6427] [Help] [answered]
### ImageMagick version

7.1.1.6

### Operating system, version and so on

Windows 11

### TLDR;
Please can someone help me modify the below command so that it will use the image's own background colour (or say the image of the top-left pixel if that can't be readily worked out) to remove the background, rather than a hard-coded colour like black/white...

`magick.exe %1 -bordercolor white -border 1x1 -alpha set -channel RGBA -fuzz 10%% -fill none -floodfill +0+0 white -shave 1x1 %1`

***

### Additional context / background
I put together this IM 1-liner from reading the guide and 1-2 posts, which automatically removes the background of an image, it works well generally, however the example script I based it on assumes a fully white image background.
`magick.exe %1 -bordercolor WHITE -border 1x1 -alpha set -channel RGBA -fuzz 10%% -fill none -floodfill +0+0 WHITE -shave 1x1 %1`

I wrote a twin script which does the same, except assuming a black background, however as you can probably imagine, and guess where I'm leading to - while this colour hard-coding covers most cases, it doesn't obviously work for any image.
`magick.exe %1 -bordercolor BLACK -border 1x1 -alpha set -channel RGBA -fuzz 10%% -fill none -floodfill +0+0 BLACK -shave 1x1 %1`

I was reading about how to test the colour of a specific pixel, and got as far as this...
`magick.exe %1 -format "%[fx:int(255*p{1,1}.r)],%[fx:int(255*p{1,1}.g)],%[fx:int(255*p{1,1}.b)]" info:` 
...and while I could load this into string into memory, try and parse it and then dynamically generate a ImageMagick command using that colour value, I thought there is probably an easier and better way / I was probably making this harder than it needed to be, but I am at the limit of my IM knowledge unfortunately.

Thanks in advance! 👍🏼

**Comment by snibgo** [comment, marked as answer]
To set the bordercolor to the same colour as the top-left pixel:
```
magick toes.png -bordercolor "%[pixel:p{0,0}]" -border 20 x.png
```
If you do this in a Windows BAT dcript, you need to double the `%`.

See https://imagemagick.org/script/fx.php

With more work, you can set the bordercolor to the mean (average) of all the pixels that are around the edge of the image. You can do this by making 4 crops, rotating 2 of these, appending them into an image with one dimension, scaling to a single pixel, and using the `"%[pixel:p{0,0}]"` of that. EDIT: See my script `edgecols.bat` on my page [Subimage rectangles](http://im.snibgo.com/subimrect.htm). This also shows a script to calculate a good `-fuzz` value.

**Reply by gaiking-uk** [reply]
Thanks @snibgo -- I incorporated your code into my script and that did the trick! 👍🏼

Full / amended code (as you rightly spotted and pointed out, intended to be run from a .bat, i.e. batch file)...
`Magick.exe "%~1" -bordercolor "%%[pixel:p{0,0}]" -border 1x1 -alpha set -channel RGBA -fuzz 20%% -fill none -floodfill +0+0 "%%[pixel:p{0,0}]" -shave 1x1 "%~dpn1.png"`

PS -- Additional tweak made to input/parameter refernce, i.e. %1 --> "%~1" and "%~dpn1.png" to allow the script to take additional filetypes, such as JPGs and convert these directly to PNGs with backgrounds removed.

**Comment by gaiking-uk** [comment]
NB: I did consider your comment / suggestion? about further improving the script by checking the edges and calculating a better a colour to use for the transparency, and while I appreciate it - I figured that I mainly wrote the script to remove the background on relatively simple images (like I say, initially it was just on ones which had a plain, all-white background). I get your point that the above script is limited and can't deal with a background that changes (or I guess more accurately, a background that changes more than the `fuzz` factor).

I did a bit of pre-exploration on this and constructed the below test images (with the results of the improved script)...
![Results](https://github.com/ImageMagick/ImageMagick/assets/37931190/1a308557-c1d6-4c25-b44e-5addda02fdfe)

On review, I appreciate your advice but think in reality the current script is probably 'good enough' for the likely use-case intended and if the background got too complex / varied, I would probably switch and use a more dedicated background removal tool for those images... If you disagree or have any other thoughts though happy to listen (that's why I took the extra time to follow-up on this sub-point, as I recognise you have a lot more knowledge in this area than I do). 👍🏼

**Comment by snibgo** [comment]
Yes, a "background" may be a solid color, or a slightly varying colour (eg a product photographed against a whitish background) or very complex (eg people photographed in front of a landscape or cityscape). So solutions vary from the very simple to very complex.

**Reply by gaiking-uk** [reply]
@snibgo -- Fair point, I suppose a "background" is a bit of a general term that can cover a range of different things really.

Have been checking out http://im.snibgo.com ... well, I say *checking out* but it is a bit above my level of IM / windows scripting TBH -- very cool though, you clearly know you stuff!

I had 1-2 other scripts and thoughts/questions that would be great to get your input on if you have some time... I wanted to send you a PM first (rather than just post and tag you publicly) and noticed at the [bottom of your homepage](https://im.snibgo.com/#:~:text=send%20a%20PM%20to%20me%20on%20that%20forum) you share you ImageMagick forum username, however [new user sign-up is blocked on the (old) forum site](https://www.imagemagick.org/discourse-server/ucp.php?mode=register) (presumably since they switched to github as the primary Q&A resource), so I can't message you there... Is there another way I can reach out / PM you?
